src/main/java/com/aws/greengrass/deployment/DeploymentConfigMerger.java

```diff
@@ -177,51 +177,77 @@ private boolean validateNucleusConfig(CompletableFuture<DeploymentResult> totall
-                if (service.getStateModTime() > mergeTime && State.BROKEN.equals(state)) {
-                    logger.atWarn(MERGE_CONFIG_EVENT_KEY).kv(SERVICE_NAME_LOG_KEY, service.getName())
-                            .log("merge-config-service BROKEN");
-                    throw new ServiceUpdateException(
-                            String.format("Service %s in broken state after deployment", service.getName()),
-                            DeploymentErrorCode.COMPONENT_BROKEN,
-                            DeploymentErrorCodeUtils.classifyComponentError(service, kernel));
-                }
-                if (!service.reachedDesiredState()) {
-                    allServicesRunning = false;
-                    continue;
-                }
-                if (State.RUNNING.equals(state) || State.FINISHED.equals(state) || !service.shouldAutoStart()
-                        && service.reachedDesiredState()) {
-                    continue;
-                }
-                allServicesRunning = false;
-            }
-            if (allServicesRunning) {
+        while (!areAllServiceInDesiredState(servicesToTrack, mergeTime, kernel)) {
+            Thread.sleep(WAIT_SVC_START_POLL_INTERVAL_MILLISEC); // hardcoded
+        }
+    }
+
+    /**
+     * Completes the provided future when all the listed services are running.
+     * Exits early if the future is cancelled
+     *
+     * @param servicesToTrack       services to track
+     * @param mergeTime             time the merge was started, used to check if a service is broken due to the merge
+     * @param kernel                kernel
+     * @param totallyCompleteFuture used to check if a deployment is cancelled
+     * @throws InterruptedException   if the thread is interrupted while waiting here
+     * @throws ServiceUpdateException if a service could not be updated
+     */
+    public static void waitForServicesToStart(Collection<GreengrassService> servicesToTrack, long mergeTime,
+                                              Kernel kernel, CompletableFuture<DeploymentResult> totallyCompleteFuture)
+            throws InterruptedException, ServiceUpdateException {
+        while (!areAllServiceInDesiredState(servicesToTrack, mergeTime, kernel)) {
+            if (totallyCompleteFuture.isCancelled()) {
+                logger.atWarn(MERGE_CONFIG_EVENT_KEY)
+                        .log("deployment cancelled while waiting for services to start");
                 return;
             }
             Thread.sleep(WAIT_SVC_START_POLL_INTERVAL_MILLISEC); // hardcoded
         }
     }
 
+    private static boolean areAllServiceInDesiredState(Collection<GreengrassService> servicesToTrack, long mergeTime,
+                                                       Kernel kernel) throws ServiceUpdateException {
+        boolean allServicesRunning = true;
+        for (GreengrassService service : servicesToTrack) {
+            State state = service.getState();
```

Comment:
Nit: What you have currently is cleaner but if we want to cancel fast (where there are a lot of services), can we do the `totallyCompleteFuture` cancelled check with each service here although there's no blocking here.

Reply:
I can make the changes in the next PR.
Currently this function is being used by both `waitForServicesToStart` one with CompletableFuture as a parameter, one without.
I plan to remove the `waitForServicesToStart` without CompletableFuture as a parameter in my next PR because it's used by Kernel Update Deployments